## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto Part One

### 1. Start by creating a new Notebook for this assignment.

In [108]:
#Library import
from bs4 import BeautifulSoup
import requests

### 2. Scrap Wikipedia Page

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

In [109]:
#Accessing Url
url=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

#BeautifulSoup
soup=BeautifulSoup(url,'lxml')
neighborhood_table=soup.find('table',{'class':'wikitable sortable'})

In [110]:
# Search for tr tag
table_rows = neighborhood_table.find_all('tr') 

#Initialize Empty list for storing row data
neighborhood_data=[] 

#Iterate through each row in table_rows and append the it to list
for tr in table_rows:
    td = tr.find_all('td')
    row = [i.text for i in td]
    neighborhood_data.append(row)   

### 3. To create the above dataframe:

- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you      will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will      be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
   If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
   Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

- Data Frame Creation

In [111]:
#Remove null and empy header
neighborhood_data.remove([ ]) 

In [112]:
#import pandas library
import pandas as pd

#Initialize Empty DataFrame with columne name 'PostalCode','Borough','Neighborhood'
df=pd.DataFrame(columns=['PostalCode','Borough','Neighborhood']) 

#Iterate through each item in list neighborhood_data and append it to data frame 
for each in neighborhood_data:
    df=df.append({'PostalCode': each[0],
                  'Borough':each[1],
                 'Neighborhood':each[2]},ignore_index=True)

Each row in data frame have new line character \n below steps is used to remove it

In [113]:
#Each row values have \n character, replace function use for removing it
#Remove new line character \n from PostalCode, Borough and Neighborhood

df.PostalCode.replace({r'\n$':''},regex=True,inplace=True)

df.Borough.replace({r'\n$':''},regex=True,inplace=True)

df.Neighborhood.replace({r'\n$':''},regex=True,inplace=True)

- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [114]:
#Deleting all row whose Borough is "Not assigned"
import numpy as np
index_to_drop=np.where(df['Borough']=='Not assigned')
df.drop(df.index[index_to_drop],inplace=True)
df.reset_index(drop=True,inplace=True)

- Display Data Frame

In [115]:
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [116]:
df.to_csv('Dataset_For_Neighborhoods_in_Toronto.csv',index=False)

- Check and Merge Different Neighborhood for Same PostalCode

In [117]:
#Get the unique PostalCode size
df.PostalCode.unique().size

103

In [118]:
#Display Total PostalCode Size
df.PostalCode.size

103

The count of Unique PostalCode is same as that of Total PostalCode (which is 103), Hence there is no duplicate PostalCode is present and need merging of Neighborhood.

In [119]:
np.where(df.PostalCode.duplicated())

(array([], dtype=int64),)

- Shape

In [120]:
df.shape

(103, 3)

##### End of Part one!!!

## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto Part Two

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:

### Load and Explore datasets

The Google Maps Geocoding API not working properly so, donloaded CSV file from given link and merge with the data frame scarped from wikipedia.

In [121]:
url='http://cocl.us/Geospatial_data'
df1=pd.read_csv(url)

In [122]:
#Rename column in dataframe df1
df1.rename(columns={'Postal Code':'PostalCode'},inplace=True)

In [123]:
df1.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [124]:
#Merge Data Frame df and df1 with common column PostalCode
data=pd.DataFrame.merge(df,df1,on='PostalCode')

In [125]:
#check the Shape of the data Frame
data.shape

(103, 5)

### Quick Display

In [126]:
data.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

### End of Part Two!!!

## Peer-graded Assignment: Segmenting & Clustering Neighborhoods in Toronto Part Three

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.
Just make sure:

- to add enough Markdown cells to explain what you decided to do and to report any observations you make.
- to generate maps to visualize your neighborhoods and how they cluster together.

In [128]:
data.to_csv('capstone_project_dataset.csv',index=False)

Get The Borough that contains word Toronto

In [129]:
toronto_data = data[data['Borough'].str.contains('Toronto')].reset_index(drop=True)
toronto_data.head(5)

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

##### Create a map of Toronto with neighborhoods superimposed on top.

In [130]:
##### import folium
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'],toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

##### Define Foursquare Credentials and Version

In [131]:
CLIENT_ID = 'SJPO45IX0J02NAO0CKCPRIXLBEC0TKBTQF1PFRG3C1IDC050' # your Foursquare ID
CLIENT_SECRET = 'PID2MUPRKRIMABIBRTKEYJN2J1C1EKSWGVQZGVUVYPFHO25S' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SJPO45IX0J02NAO0CKCPRIXLBEC0TKBTQF1PFRG3C1IDC050
CLIENT_SECRET:PID2MUPRKRIMABIBRTKEYJN2J1C1EKSWGVQZGVUVYPFHO25S


##### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [132]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.

In [133]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


##### Now, let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters.

In [134]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=SJPO45IX0J02NAO0CKCPRIXLBEC0TKBTQF1PFRG3C1IDC050&client_secret=PID2MUPRKRIMABIBRTKEYJN2J1C1EKSWGVQZGVUVYPFHO25S&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [135]:
results = requests.get(url).json()
#results

### Explore Neighborhoods in Toronto

##### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [136]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Create new dataframe called *toronto_venues*.

In [137]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West,  Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport


Let's check how many venues were returned for each neighborhood

In [138]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     19   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             32   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          26   
India Bazaar, The Beaches West                                         21   
Kensington Market, Chinatown, Grange Park                              68   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               47   
Moore Park, Summerhill East                                             1   
North Toronto West,  Lawrence Park                                     20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     35   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         98   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             5   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      19   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              32   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [139]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 233 uniques categories.


### Analyse Neighborhoods in Toronto

In [140]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...                 0                 0      0   
1             0  ...                 0                 0      0   
2             0  ...                 0                 0      0   
3             0  ...                 0                 0      0   
4             0  ...                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  
0                      0         0          0              0  
1                      0         0          0              0  
2                      0         0          0              0  
3                      0         0          0              0  
4                      0         0          0              0  

[5 rows x 233 columns]

##### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [141]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
#toronto_grouped

First, let's write a function to sort the venues in descending order.

In [142]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [143]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...      Airport Terminal   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café           Cheese Shop        Farmers Market   
1        Breakfast Spot           Coffee Shop          Climbing Gym   
2                  Park         Auto Workshop            Comic Shop   
3        Airport Lounge       Airport Service                 Plane   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant            Restaurant   
1         Burrito Place    Italian Restaurant            Restaurant   
2           Pizza Place               Butcher      Recording Studio   
3       Harbor / Marina   Rental Car Location           Coffee Shop   
4   Japanese Restaurant           Salad Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar          Cocktail Bar                 Lounge  
1          Intersection               Stadium                    Bar  
2         Burrito Place            Restaurant                Brewery  
3         Boat or Ferry                   Bar           Airport Gate  
4          Burger Joint                   Bar       Department Store

## Cluster Neighborhoods in Toronto

Run *k*-means to cluster the neighborhood into 5 clusters.

In [144]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [145]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.676357 -79.293031               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                   Pub                Bakery   
1                 Diner           Yoga Studio    College Auditorium   
2           Coffee Shop       Bubble Tea Shop   Japanese Restaurant   
3           Coffee Shop        Clothing Store   American Restaurant   
4                   Pub                 Trail     Health Food Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café               Theater        Breakfast Spot   
1              Beer Bar         Smoothie Shop        Sandwich Place   
2                  Café    Italian Restaurant        Cosmetics Shop   
3        Cosmetics Shop          Cocktail Bar            Restaurant   
4          Dessert Shop          Cupcake Shop          Dance Studio   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio            Shoe Store                  Hotel  
1         Burrito Place                  Café  General Entertainment  
2                 Hotel              Tea Room       Ramen Restaurant  
3    Seafood Restaurant      Department Store                    Gym  
4         Deli / Bodega      Department Store                  Diner

Visualize the resulting clusters

In [146]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

##### End Of Part Three and Notebook!!!